### PR TITLE
expression: add year buildin function

### PR DESCRIPTION
**Pull request by Kingwl**
## What have you changed? (mandatory)
add year function

Please **NOTE** that:
- Do not assume reviewers understand the original issue
#3275

## What are the type of the changes? (mandatory)

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 1]
```diff
@@ -85,6 +85,31 @@ impl ScalarFunc {
...
+    #[inline]
+    pub fn year<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<i64>> {
+        let t = match self.children[0].eval_time(ctx, row) {
+            Err(err) => return Error::handle_invalid_time_error(ctx, err).map(|_| None),
+            Ok(None) => {
+                return Error::handle_invalid_time_error(
```
Comment (spongedu) [inline src/coprocessor/dag/expr/builtin_time.rs]: I think there's no need calling `Error::handle_invalid_time_error`here . we can return ok(None) directly.
```sql
mysql> select year(null);
+------------+
| year(null) |
+------------+
|       NULL |
+------------+
1 row in set (0.00 sec)
```

No error or warning when input is `NULL`

**Comment by Kingwl** [comment]
not my fault

**Comment by hicqu** [comment]
LGTM.

**Comment by Kingwl** [comment]
should i add the `Mode` flag in this pr?

**Comment by spongedu** [comment]
@Kingwl I suggest so. builtin `year` seems need this

**Comment by huachaohuang** [comment]
@spongedu Thanks for your review too :)

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 2]
```diff
@@ -311,4 +331,52 @@ mod test {
...
+        }
+        // test NULL case
+        let input = datum_expr(Datum::Null);
+        let f = scalar_func_expr(ScalarFuncSig::Year, &[input]);
+        let op = Expression::build(&mut ctx, f).unwrap();
+        let got = op.eval(&mut ctx, &[]).unwrap();
+        assert_eq!(got, Datum::Null);
+
+        // test zero case
+        let cfg = EvalConfig::new()
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_time.rs]: Oh, `EvalConfig` is updated, you need to do it like this now: https://github.com/tikv/tikv/pull/3622/files#diff-1377468ba7a8bb8132fa4d0c4f429700L281

**Pull request merged**